pkg/bucketaccess/bucketaccess_controller.go

```diff
@@ -91,7 +95,10 @@ func (bal *BucketAccessListener) Add(ctx context.Context, inputBucketAccess *v1a
 	}
 
 	bucketAccessClass, err := bal.bucketAccessClasses().Get(ctx, bucketAccessClassName, metav1.GetOptions{})
-	if err != nil {
+	if kubeerrors.IsNotFound(err) {
+		bal.recordEvent(inputBucketAccess, v1.EventTypeWarning, events.GrantingAccessFailed, "BucketAccessClass provided in the BucketAccess does not exist")
```

Comment:
most controllers I've seen, when they encounter an issue, the error encountered is included with the event reported. I think we should do this. 

What if we used a process like:
```go
err := errors.Wrapf(err, "BucketAccessClass %q provided in BucketAccess %q does not exist", bucketAccessClassName, BucketAccess.NamespacedName())
b.recordEvent(inputBucketAccess, v1.EventTypeWarning, events.GrantingAccessFailed, err.Error())
return err
```

I could also suggest that we could make the process a bit more reliable and extensible by using a pattern we have also used to help us reliably emit events in Rook. We immediately call a do-er function that returns some info about every failure. For every single failure, we emit an event as well as a log message about it.

```
func (bal *BucketAccessListener) Add(ctx context.Context, inputBucketAccess *v1alpha1.BucketAccess) error {
    errReason, err := doAdd(ctx context.Context, inputBucketAccess *v1alpha1.BucketAccess)
    if err != nil {
        outErr := errors.Wrapf(err, "failed to add BucketAccess %q", inputBucketAccess.NamespacedName())
        klog.ErrorS(outErr)
        bal.recordEvent(inputBucketAccess, v1.EventTypeWarning, errReason, outErr.Error())
        
    }
}


func doAdd(ctx context.Context, inputBucketAccess *v1alpha1.BucketAccess) (string, error) {
    // the code currently in Add() goes here
}
```


There is an additional piece of this: when dealing with k8s resources, you MUST report resources that are namespace by both their namespace and name because "name" can collide across namespaces in the cluster. Any log/error messages will be hard to debug if there are multiples in a cluster. It should be standard practice to return the namespaced name of resources being reconciled, and the easiest way is to turn the `Reconcile()` call into a basic wrapper for a `doReconcile()` helper; that way all errors will at minimum get the reconciled resource's namespacedName attached.

Something noteworthy is that `klog` supports structured logging, though I'm more familiar with `logr`, meaning you can attach info (like `klog.AddS("nsName", obj.NamespacedName())`) to the logger and pass that to children, who will then output the nsName with all their logs, indicating in each log what resource it corresponds to.

Changing the loggging would require changing quite a bit of the controller code, so obviously that can't be done here. But since this is a library where so many people may come to contribute, it might be extremely valuable to set code in place that makes sure important info needed for debugging (especially namespaced names of resources) doesn't get forgotten. It is extremely easy to forget that and then have a hard time debugging in the future. @wlan0 thoughts?